dags/iop/parser.py

```diff
@@ -106,6 +117,51 @@ def _get_extracted_arxiv_eprint_value(self, article):
         except AttributeError:
             self.logger.error("No arXiv eprints found", dois=self.dois)
 
+    def _get_date_element(self, article):
+        try:
+            date = (
+                article.find("front/article-meta/history/date[@date-type='published']")
+                or article.find("front/article-meta/history/date[@date-type='epub']")
+                or article.find("front/article-meta/history/pub-date[@pub-type='ppub']")
+                or article.find("front/article-meta/pub-date")
+            )
+            return date
+        except AttributeError:
+            return
+
+    def _get_date_published(self, article):
+        date_element = self._get_date_element(article)
+        return self._get_date(date_element)
+
+    def _get_date(self, date):
+        day = None
+        month = None
+        try:
+            year = int(date.find("year").text)
+            self.year = year
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find year of date_published in XML")
+        try:
+            month = int(date.find("month").text)
+        # was discussed with Anne: if the day is not present, we return year and month
+        # if the month value is not present - we return just year
+        # if the year is not present - the parsing should crash, because according schema
+        # the publication_year is required, which is take from date_published
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find month of date_published in XML")
+        try:
+            day = int(date.find("day").text)
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find day of date_published in XML")
+        try:
+            date = PartialDate(year=self.year, month=month, day=day)
+            return PartialDate.dumps(date)
+        except (AttributeError, ValueError, TypeError):
```

Comment:
why do you need those 3 exceptions here? `ParialDate` on init raises `ValueError` and `self.year` is defined, isn't it?

Reply:
Exceptions: I want to identify which part of the date is missing:  day, month, or year.

No, it doesn't raise an error, because the parser functions are run after the __init__. 
I set a year in __init__, because as we discussed before, I need to reuse the year value in the journal_year field